engine/storage/datamotion/src/org/apache/cloudstack/storage/motion/StorageSystemDataMotionStrategy.java

```diff
@@ -172,78 +232,211 @@ private void validate(SnapshotInfo snapshotInfo) {
+    private boolean usingBackendSnapshotFor(SnapshotInfo snapshotInfo) {
+        String property = getProperty(snapshotInfo.getId(), "takeSnapshot");
+
+        return Boolean.parseBoolean(property);
+    }
+
+    private void handleCreateTemplateFromSnapshot(SnapshotInfo snapshotInfo, TemplateInfo templateInfo, AsyncCompletionCallback<CopyCommandResult> callback) {
         try {
             snapshotInfo.processEvent(Event.CopyingRequested);
         }
         catch (Exception ex) {
             throw new CloudRuntimeException("This snapshot is not currently in a state where it can be used to create a template.");
         }
 
-        HostVO hostVO = getHost(snapshotInfo.getDataStore().getId());
-        DataStore srcDataStore = snapshotInfo.getDataStore();
-
-        String value = _configDao.getValue(Config.PrimaryStorageDownloadWait.toString());
-        int primaryStorageDownloadWait = NumbersUtil.parseInt(value, Integer.parseInt(Config.PrimaryStorageDownloadWait.getDefaultValue()));
-        CopyCommand copyCommand = new CopyCommand(snapshotInfo.getTO(), templateInfo.getTO(), primaryStorageDownloadWait, VirtualMachineManager.ExecuteInSequence.value());
+        HostVO hostVO = getHost(snapshotInfo);
 
-        String errMsg = null;
+        boolean usingBackendSnapshot = usingBackendSnapshotFor(snapshotInfo);
+        boolean computeClusterSupportsResign = computeClusterSupportsResign(hostVO.getClusterId());
 
-        CopyCmdAnswer copyCmdAnswer = null;
+        if (usingBackendSnapshot && !computeClusterSupportsResign) {
+            throw new CloudRuntimeException("Unable to locate an applicable host with which to perform a resignature operation");
+        }
 
         try {
-            _volumeService.grantAccess(snapshotInfo, hostVO, srcDataStore);
+            if (usingBackendSnapshot) {
+                createVolumeFromSnapshot(hostVO, snapshotInfo, true);
+            }
 
-            Map<String, String> srcDetails = getSnapshotDetails(_storagePoolDao.findById(srcDataStore.getId()), snapshotInfo);
+            DataStore srcDataStore = snapshotInfo.getDataStore();
 
-            copyCommand.setOptions(srcDetails);
+            String value = _configDao.getValue(Config.PrimaryStorageDownloadWait.toString());
+            int primaryStorageDownloadWait = NumbersUtil.parseInt(value, Integer.parseInt(Config.PrimaryStorageDownloadWait.getDefaultValue()));
+            CopyCommand copyCommand = new CopyCommand(snapshotInfo.getTO(), templateInfo.getTO(), primaryStorageDownloadWait, VirtualMachineManager.ExecuteInSequence.value());
+
+            String errMsg = null;
+
+            CopyCmdAnswer copyCmdAnswer = null;
 
-            copyCmdAnswer = (CopyCmdAnswer)_agentMgr.send(hostVO.getId(), copyCommand);
-        }
-        catch (Exception ex) {
-            throw new CloudRuntimeException(ex.getMessage());
-        }
-        finally {
             try {
```

Comment:
I believe we can extract lines 273-288 to a method. Then, that block comment there can become a java doc. It would also enable the writing of a test case.